src/plugins/ui_core_views/evaluation.ts

```diff
@@ -253,33 +510,121 @@ export class EvaluationPlugin extends UIPlugin {
         cellData.dependencies,
         ...compilationParameters
       );
-      cellsBeingComputed.delete(cellId);
-      if (Array.isArray(computedCell.value)) {
-        // if a value returns an array (like =A1:A3)
-        throw new Error(_lt("This formula depends on invalid values"));
+
+      if (!isMatrix(computedValue)) {
+        if (isMatrix(computedFormat)) {
+          throw "A format matrix should never be associated with a scalar value";
+        }
+        return createEvaluatedCell(computedValue, cellData.format || computedFormat);
+      }
+
+      let formatFromPosition: (i: number, j: number) => string | undefined;
+
+      if (isMatrix(computedFormat)) {
+        formatFromPosition = (i, j) => computedFormat[i][j];
+        const sameDimensions =
+          computedValue.length === computedFormat.length &&
+          computedValue[0].length === computedFormat[0].length;
+        if (!sameDimensions) {
+          throw "Formats and values should have the same dimensions !!!";
+        }
+      } else {
+        formatFromPosition = (i, j) => computedFormat;
+      }
+
+      const { sheetId, col, row } = this.getters.getCellPosition(cellId);
+      const parentXC = cellPositionToXc({ sheetId, col, row });
+
+      const numberOfCols = this.getters.getNumberCols(sheetId);
+      const numberOfRows = this.getters.getNumberRows(sheetId);
+      const enoughCols = col + computedValue.length <= numberOfCols;
+      const enoughRows = row + computedValue[0].length <= numberOfRows;
+      if (!enoughCols || !enoughRows) {
+        if (enoughCols) {
+          throw "Result couldn't be automatically expanded. Please insert more rows.";
+        }
+        if (enoughRows) {
+          throw "Result couldn't be automatically expanded. Please insert more columns.";
+        }
+        throw "Result couldn't be automatically expanded. Please insert more columns and rows.";
       }
-      return createEvaluatedCell(computedCell.value, cellData.format || computedCell.format);
+
+      mapSpreadPositionInMatrix(computedValue, updateSpreadCandidates);
+      mapSpreadPositionInMatrix(computedValue, checkCollision);
+      mapSpreadPositionInMatrix(computedValue, spreadValues);
+      return createEvaluatedCell(computedValue[0][0], cellData.format || formatFromPosition(0, 0));
     };
 
-    const compilationParameters = this.getCompilationParameters((cell) => computeCell(cell)());
+    const compilationParameters = this.getCompilationParameters(computeCell);
 
-    for (const cell of this.getAllCells()) {
-      this.setEvaluatedCell(cell.id, computeCell(cell));
+    for (let currentCycle = 0; currentCycle < this.maxCycle; ++currentCycle) {
+      this.currentXcsToUpdate = new Set([...this.currentXcsToUpdate, ...this.nextXcsToUpdate]);
+      for (const xc of this.currentXcsToUpdate) {
+        for (const child of this.potentialSpreaders.getParents(xc)) {
+          this.potentialSpreaders.removeEdge(child.xc, xc);
+        }
+      }
```

Comment:
Is this equivalent ? It looks like all tests pass with this change
```suggestion
      for (const xc of this.currentXcsToUpdate) {
        this.potentialSpreaders.removeNode(xc);
      }
```

Reply:
change done